compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToUKernels.cpp

```diff
@@ -0,0 +1,296 @@
+    // Here the matmul ukernel op won't read the existing accumulator, so its
+    // defining op can be discarded.
+    if (auto fillOp = out.getDefiningOp<linalg::FillOp>()) {
+      out = fillOp.getDpsInitOperand(0)->get();
+    }
+  }
+
+  Location loc = op.getLoc();
+  // Since this is invokved after the final pack-and-transpose we need to
+  // extract dimensions for M*K*k*m*m0*k0 x K*N*n*k*k0*n0 -> M*N*n*m*m0*n0.
+  Value M = rewriter.create<tensor::DimOp>(loc, lhs, 0);
+  Value N = rewriter.create<tensor::DimOp>(loc, rhs, 1);
+  Value K = rewriter.create<tensor::DimOp>(loc, lhs, 1);
+  Value m = rewriter.create<tensor::DimOp>(loc, lhs, 3);
+  Value n = rewriter.create<tensor::DimOp>(loc, rhs, 2);
+  Value k = rewriter.create<tensor::DimOp>(loc, lhs, 2);
+
+  auto getDimAsI32 = [](RewriterBase &rewriter, Location loc, Value value,
+                        int dim) -> Value {
+    return rewriter.create<arith::IndexCastOp>(
+        loc, rewriter.getI32Type(),
+        rewriter.create<tensor::DimOp>(loc, value, dim));
+  };
+  Value m0 = getDimAsI32(rewriter, loc, lhs, 4);
+  Value n0 = getDimAsI32(rewriter, loc, rhs, 5);
+  Value k0 = getDimAsI32(rewriter, loc, lhs, 5);
+  auto fn = getFnNameAndDefAttrs(ukernelName, rewriter, targetAttr);
+
+  // Create UKernel for AMD-AIE.
+  // TODO(avarma): I have gotten rid of the `flag` for now. Will address this as
+  //               things pertaining to microkernel gets clearer.
+  auto genericMicroKernelOp = rewriter.create<IREE::Codegen::UKernelGenericOp>(
+      loc, outType, fn.name, ValueRange{lhs, rhs}, out,
+      ValueRange{M, N, K, m, n, k, m0, n0, k0},
+      /*fn_def_attrs=*/rewriter.getDictionaryAttr(fn.defAttrs),
+      /*strided_outer_dims=*/rewriter.getIndexAttr(1));
+  return cast<IREE::Codegen::UKernelOpInterface>(
+      genericMicroKernelOp.getOperation());
+}
+
+using TargetPredicate = std::function<bool(IREE::HAL::ExecutableTargetAttr)>;
+
+/// TODO(avarma): Currently since we get linalg.generic in the innermost loop
+///               via simple-pack pipeline, the following function is fixating
+///               on the same. Make it templatized once we support more ops via
+///               microkernel.
+struct LowerToUKernelPattern : OpRewritePattern<linalg::GenericOp> {
+  LowerToUKernelPattern(MLIRContext *context, TargetPredicate targetPredicate)
+      : OpRewritePattern<linalg::GenericOp>(context),
+        targetPredicate(targetPredicate) {}
+
+  LogicalResult matchAndRewrite(linalg::GenericOp op,
+                                PatternRewriter &rewriter) const override {
+    if (targetPredicate &&
+        !targetPredicate(IREE::HAL::ExecutableTargetAttr::lookup(op))) {
+      return failure();
+    }
+
+    FailureOr<IREE::Codegen::UKernelOpInterface> ukernelOp;
+    if (isMatmul(op)) {
+      ukernelOp = matchDAGForUKernel(rewriter, op, "matmul");
+    } else {
+      // TODO(avarma): This looks murky. There are dispatches with fused
```

Comment:
This is OK actually. If we want to use a ukernel + fusion, we will need an intermediate stack buffer allocation in L1 to satisfy dependencies. So you can drop the TODO I think